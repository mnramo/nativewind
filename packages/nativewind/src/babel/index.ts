import { resolve, sep, posix } from "node:path";
import { statSync } from "node:fs";

import micromatch from "micromatch";

import type { Config } from "tailwindcss";
import resolveConfigPath from "tailwindcss/lib/util/resolveConfigPath";
import resolveConfig from "tailwindcss/resolveConfig";
import { validateConfig } from "tailwindcss/lib/util/validateConfig";

import type { ConfigAPI, PluginPass, Visitor } from "@babel/core";
import { addNamed } from "@babel/helper-module-imports";

import { extractStyles } from "../postcss/extract-styles";
import { TailwindcssReactNativeBabelOptions } from "./types";
import { visitor, VisitorState } from "./visitor";

import { nativePlugin } from "../tailwind/native";
import { expressionStatement } from "@babel/types";

export default function (
  api: ConfigAPI,
  options: TailwindcssReactNativeBabelOptions,
  cwd: string
) {
  /**
   * Get the users config
   */
  const userConfigPath = resolveConfigPath(
    options.tailwindConfig || options.tailwindConfigPath
  );

  let tailwindConfig: Config;

  if (userConfigPath === null) {
    tailwindConfig = resolveConfig({
      ...options.tailwindConfig,
      plugins: [
        nativePlugin(options),
        ...(options?.tailwindConfig?.plugins ?? []),
      ],
    });
  } else {
    api.cache.using(() => statSync(userConfigPath).mtimeMs);
    // eslint-disable-next-line  @typescript-eslint/no-explicit-any
    (api as any).addExternalDependency(userConfigPath);

    // eslint-disable-next-line @typescript-eslint/no-var-requires,unicorn/prefer-module
    const userConfig = require(userConfigPath);

    const newConfig = resolveConfig({
      ...userConfig,
      plugins: [nativePlugin(options), ...(userConfig?.plugins ?? [])],
    });

    tailwindConfig = validateConfig(newConfig);
  }

  /**
   * Resolve their content paths
   */
  const contentFilePaths = (
    Array.isArray(tailwindConfig.content)
      ? tailwindConfig.content.filter(
          (filePath): filePath is string => typeof filePath === "string"
        )
      : tailwindConfig.content.files.filter(
          (filePath): filePath is string => typeof filePath === "string"
        )
  ).map((contentFilePath) => normalizePath(resolve(cwd, contentFilePath)));

  const allowModuleTransform = Array.isArray(options.allowModuleTransform)
    ? ["react-native", "react-native-web", ...options.allowModuleTransform]
    : "*";

  const programVisitor: Visitor<
    PluginPass & {
      opts: TailwindcssReactNativeBabelOptions;
    }
  > = {
    Program: {
      enter(projectPath, state) {
        const filename = state.filename;

        if (!filename) return;

        const isInContent = micromatch.isMatch(
          normalizePath(filename),
          contentFilePaths
        );

        if (!isInContent) {
          return;
        }

        let canCompile = true;
        let canTransform = true;

        switch (state.opts.mode) {
          case "compileOnly": {
            canTransform = false;
            break;
          }
          case "transformOnly": {
            canCompile = false;
            break;
          }
        }

        const visitorState: VisitorState = {
          ...state,
          rem: 16,
          didTransform: false,
          canCompile,
          canTransform,
          filename,
          allowModuleTransform,
          allowRelativeModules: contentFilePaths,
          blockList: new Set(),
          tailwindConfig: tailwindConfig,
        };

        projectPath.traverse(visitor, visitorState);

        const { didTransform } = visitorState;

        if (didTransform) {
          addNamed(projectPath, "StyledComponent", "nativewind");
        }

        const output = extractStyles({
          ...tailwindConfig,
          content: [filename],
          // If the file doesn't have any Tailwind styles, it will print a warning
          // We force an empty style to prevent this
          safelist: tailwindConfig.safelist ?? ["babel-empty"],
        });

        if (!output.hasStyles) return;

        projectPath.pushContainer(
          "body",
          expressionStatement(output.stylesheetCreateExpression)
        );

        addNamed(projectPath, "NativeWindStyleSheet", "nativewind");
      },
    },
  };

  return {
    visitor: programVisitor,
  };
}

function normalizePath(filePath: string) {
  /**
   * This is my naive way to get path matching working on Windows.
   * Basically I turn it into a posix path which seems to work fine
   *
   * If you are a windows user and understand micromatch, can you please send a PR
   * to do this the proper way
   */
  return filePath.split(sep).join(posix.sep);
}
