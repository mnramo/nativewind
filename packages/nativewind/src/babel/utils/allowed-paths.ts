import micromatch from "micromatch";
import { resolve, sep, posix } from "node:path";
import { Config } from "tailwindcss";
import { TailwindcssReactNativeBabelOptions, AllowPathOptions } from "../types";

const defaultContent: NonNullable<Config["content"]> = {
  files: ["*"],
  extract: undefined,
  transform: undefined,
};

interface GetAllowedOptionsOptions {
  allowModuleTransform: AllowPathOptions;
  allowRelativeModules: AllowPathOptions;
}

export function getAllowedOptions(
  { content = defaultContent }: Config,
  { allowModuleTransform = "*" }: TailwindcssReactNativeBabelOptions
): GetAllowedOptionsOptions {
  const contentPaths = (
    Array.isArray(content) ? content : content.files
  ).filter((a): a is string => typeof a === "string");

  return {
    allowModuleTransform: Array.isArray(allowModuleTransform)
      ? ["react-native", "react-native-web", ...allowModuleTransform]
      : allowModuleTransform,
    allowRelativeModules: contentPaths.length === 0 ? "*" : contentPaths,
  };
}

export interface IsAllowedProgramPathOptions {
  path: string;
  allowRelativeModules: AllowPathOptions;
  cwd: string;
}

export function isAllowedProgramPath({
  path,
  allowRelativeModules,
  cwd,
}: IsAllowedProgramPathOptions) {
  if (allowRelativeModules === "*") {
    return true;
  }

  /**
   * This is my naive way to get path matching working on Windows.
   * Basically I turn it into a posix path which seems to work fine
   *
   * If you are a windows user and understand micromatch, can you please send a PR
   * to do this the proper way
   */
  const posixPath = path.split(sep).join(posix.sep);

  return allowRelativeModules.some((modulePath) => {
    return micromatch.isMatch(
      posixPath,
      resolve(cwd, modulePath).split(sep).join(posix.sep)
    );
  });
}
