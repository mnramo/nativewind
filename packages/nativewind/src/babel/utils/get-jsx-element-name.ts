import { types } from "@babel/core";
import {
  isJSXIdentifier,
  isJSXNamespacedName,
  JSXIdentifier,
  JSXMemberExpression,
  JSXNamespacedName,
} from "@babel/types";

/**
 * Get the name of a JXS Element
 *
 * @remarks
 *
 * Valid JSX:
 *
 * <MyComponent />
 * <A.MyComponent />
 * <B.A.MyComponent />
 * <C.B.A.MyComponent />
 * ..etc
 *
 * @privateRemarks
 *
 * This was abstracted as it needs to be recursive
 *
 */
export function getJSXElementName(node: types.JSXOpeningElement) {
  return getElementName(node.name);
}

/**
 * Recursive helper function for getJSXElementName
 */
function getElementName(
  node: JSXIdentifier | JSXNamespacedName | JSXMemberExpression
): string {
  if (isJSXIdentifier(node)) {
    return node.name;
  } else if (isJSXNamespacedName(node)) {
    return node.name.name;
  } else {
    return getElementName(node.object);
  }
}
