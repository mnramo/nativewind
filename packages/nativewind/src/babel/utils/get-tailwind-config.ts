import { existsSync } from "node:fs";
import { Config } from "tailwindcss";
import resolveTailwindConfig from "tailwindcss/resolveConfig";

import { nativePlugin, NativePluginOptions } from "../../tailwind/native";

export interface GetTailwindConfigOptions extends NativePluginOptions {
  tailwindConfigPath?: string;
}

export function getTailwindConfig(
  fullConfigPath: string,
  options: GetTailwindConfigOptions
): Config {
  const { tailwindConfigPath } = options;

  let userConfig: Partial<Config> = {};
  if (existsSync(fullConfigPath)) {
    // eslint-disable-next-line unicorn/prefer-module
    userConfig = require(fullConfigPath);
  } else if (tailwindConfigPath) {
    throw new Error(`Unable to find config ${fullConfigPath}`);
  }

  const mergedConfig = {
    ...userConfig,
    plugins: [nativePlugin(options), ...(userConfig.plugins ?? [])],
  } as Config;

  return resolveTailwindConfig(mergedConfig);
}
