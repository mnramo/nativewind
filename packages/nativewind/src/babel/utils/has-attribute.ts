import { NodePath } from "@babel/core";
import {
  isJSXIdentifier,
  isJSXSpreadAttribute,
  JSXElement,
} from "@babel/types";

export function someAttributes(path: NodePath<JSXElement>, names: string[]) {
  const openingElement = path.node.openingElement;

  return openingElement.attributes.some((attribute) => {
    // Ignore spreads, we cannot process them
    if (isJSXSpreadAttribute(attribute)) {
      return false;
    }

    return names.some((name) => isJSXIdentifier(attribute.name, { name }));
  });
}
