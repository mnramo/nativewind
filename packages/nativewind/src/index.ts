export * from "./styled";
export * from "./styled-component";
export * from "./styled/use-tailwind";
export { NativeWindStyleSheet } from "./style-sheet";
export { useColorScheme } from "./use-color-scheme";
export { platformSelect } from "./utils/platform-select";
export { withExpoSnack } from "./expo-snack";
export type { StyledProps } from "./utils/styled";
