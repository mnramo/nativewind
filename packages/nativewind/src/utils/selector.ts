import type { PlatformOSType } from "react-native";

/* prettier-ignore */ export const PARENT_FOCUS     = 0b1000000000000000000000;
/* prettier-ignore */ export const PARENT_HOVER     = 0b0100000000000000000000;
/* prettier-ignore */ export const PARENT_ACTIVE    = 0b0010000000000000000000;
/* prettier-ignore */ export const ISO_GROUP_FOCUS  = 0b0001000000000000000000;
/* prettier-ignore */ export const ISO_GROUP_HOVER  = 0b0000100000000000000000;
/* prettier-ignore */ export const ISO_GROUP_ACTIVE = 0b0000010000000000000000;
/* prettier-ignore */ export const GROUP_FOCUS      = 0b0000001000000000000000;
/* prettier-ignore */ export const GROUP_HOVER      = 0b0000000100000000000000;
/* prettier-ignore */ export const GROUP_ACTIVE     = 0b0000000010000000000000;
/* prettier-ignore */ export const PARENT           = 0b0000000001000000000000;
/* prettier-ignore */ export const GROUP_ISO        = 0b0000000000100000000000;
/* prettier-ignore */ export const GROUP            = 0b0000000000010000000000;
/* prettier-ignore */ export const FOCUS            = 0b0000000000001000000000;
/* prettier-ignore */ export const HOVER            = 0b0000000000000100000000;
/* prettier-ignore */ export const ACTIVE           = 0b0000000000000010000000;
/* prettier-ignore */ export const RTL              = 0b0000000000000001000000;
/* prettier-ignore */ export const OSX              = 0b0000000000000000100000;
/* prettier-ignore */ export const WINDOWS          = 0b0000000000000000010000;
/* prettier-ignore */ export const WEB              = 0b0000000000000000001000;
/* prettier-ignore */ export const ANDROID          = 0b0000000000000000000100;
/* prettier-ignore */ export const IOS              = 0b0000000000000000000010;
/* prettier-ignore */ export const DARK_MODE        = 0b0000000000000000000001;

const makePseudoClassTest = (pseudoClass: string) => {
  const regex = new RegExp(`\\S+::${pseudoClass}(:|\\s|$)`);
  return regex.test.bind(regex);
};

const hasHover = makePseudoClassTest("hover");
const hasActive = makePseudoClassTest("active");
const hasFocus = makePseudoClassTest("focus");
const hasGroupHover = makePseudoClassTest("group-hover");
const hasGroupActive = makePseudoClassTest("group-active");
const hasGroupFocus = makePseudoClassTest("group-focus");
const hasGroupIsolate = RegExp.prototype.test.bind(
  /(:|\s|^)group-isolate(:|\s|^)/gi
);
const hasGroupIsolateHover = makePseudoClassTest("group-isolate-hover");
const hasGroupIsolateActive = makePseudoClassTest("group-isolate-active");
const hasGroupIsolateFocus = makePseudoClassTest("group-isolate-focus");
const hasParent = RegExp.prototype.test.bind(/(:|\s|^)parent(:|\s|$)/gi);

const hasParentHover = makePseudoClassTest("parent-hover");
const hasParentActive = makePseudoClassTest("parent-active");
const hasParentFocus = makePseudoClassTest("parent-focus");
const hasIos = makePseudoClassTest("ios");
const hasAndroid = makePseudoClassTest("android");
const hasWindows = makePseudoClassTest("windows");
const hasMacos = makePseudoClassTest("macos");
const hasWeb = makePseudoClassTest("web");
export const hasDarkPseudoClass = makePseudoClassTest("dark");

export function normalizeCssSelector(selector: string) {
  selector = selector.trim().replace(/^\.|\\/g, "");
  selector = selector.split("::")[0];
  selector = selector.split(" ").pop() as string;

  return selector;
}

export interface StateBitOptions {
  darkMode?: boolean;
  hover?: boolean;
  active?: boolean;
  focus?: boolean;
  group?: boolean;
  groupHover?: boolean;
  groupActive?: boolean;
  groupFocus?: boolean;
  isolateGroup?: boolean;
  isolateGroupHover?: boolean;
  isolateGroupActive?: boolean;
  isolateGroupFocus?: boolean;
  parent?: boolean;
  parentHover?: boolean;
  parentFocus?: boolean;
  parentActive?: boolean;
  platform?: PlatformOSType;
  rtl?: boolean;

  // Used By expo-snack
  baseBit?: number;
}

export function getStateBit(options: StateBitOptions = {}) {
  let finalBit = options.baseBit || 0;

  if (options.darkMode) finalBit |= DARK_MODE;
  if (options.platform === "ios") finalBit |= IOS;
  if (options.platform === "android") finalBit |= ANDROID;
  if (options.platform === "web") finalBit |= WEB;
  if (options.platform === "windows") finalBit |= WINDOWS;
  if (options.platform === "macos") finalBit |= OSX;
  if (options.rtl) finalBit |= RTL;
  if (options.active) finalBit |= ACTIVE;
  if (options.hover) finalBit |= HOVER;
  if (options.focus) finalBit |= FOCUS;
  if (options.group) finalBit |= GROUP;
  if (options.groupActive) finalBit |= GROUP_ACTIVE;
  if (options.groupHover) finalBit |= GROUP_HOVER;
  if (options.groupFocus) finalBit |= GROUP_FOCUS;
  if (options.isolateGroup) finalBit |= GROUP_ISO;
  if (options.isolateGroupActive) finalBit |= ISO_GROUP_ACTIVE;
  if (options.isolateGroupHover) finalBit |= ISO_GROUP_HOVER;
  if (options.isolateGroupFocus) finalBit |= ISO_GROUP_FOCUS;
  if (options.parent) finalBit |= PARENT;
  if (options.parentActive) finalBit |= PARENT_ACTIVE;
  if (options.parentHover) finalBit |= PARENT_HOVER;
  if (options.parentFocus) finalBit |= PARENT_FOCUS;

  return finalBit;
}

export function createAtRuleSelector(className: string, atRuleIndex: number) {
  return `${className}@${atRuleIndex}`;
}

export function matchesMask(value: number, mask: number): boolean {
  return (value & mask) === mask;
}

export function getSelectorMask(selector: string, rtl = false): number {
  return getStateBit({
    rtl,
    hover: hasHover(selector),
    active: hasActive(selector),
    focus: hasFocus(selector),
    groupHover: hasGroupHover(selector),
    groupActive: hasGroupActive(selector),
    groupFocus: hasGroupFocus(selector),
    isolateGroup: hasGroupIsolate(selector),
    isolateGroupHover: hasGroupIsolateHover(selector),
    isolateGroupActive: hasGroupIsolateActive(selector),
    isolateGroupFocus: hasGroupIsolateFocus(selector),
    parent: hasParent(selector),
    parentHover: hasParentHover(selector),
    parentActive: hasParentActive(selector),
    parentFocus: hasParentFocus(selector),
    darkMode: hasDarkPseudoClass(selector),
    platform: hasIos(selector)
      ? "ios"
      : hasAndroid(selector)
      ? "android"
      : hasWindows(selector)
      ? "windows"
      : hasMacos(selector)
      ? "macos"
      : hasWeb(selector)
      ? "web"
      : undefined,
  });
}
