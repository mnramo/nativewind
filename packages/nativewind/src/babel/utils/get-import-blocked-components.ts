import { createRequire } from "node:module";
import { join, dirname, basename, resolve, sep, posix } from "node:path";
import { readdirSync, lstatSync, existsSync } from "node:fs";
import micromatch from "micromatch";

import { VisitorState } from "../visitor";
import { platforms } from "../../utils/platforms";
import type { NodePath, types } from "@babel/core";

const allowedIndexFiles: string[] = [];
for (const platform of platforms) {
  for (const extension of ["js", "jsx", "ts", "tsx"]) {
    allowedIndexFiles.push(`index.${platform}.${extension}`);
  }
}

export function getImportBlockedComponents(
  path: NodePath<types.ImportDeclaration>,
  state: VisitorState
): string[] {
  const {
    filename,
    cwd,
    allowRelativeModules,
    allowModuleTransform,
    opts: { blockModuleTransform = [] },
  } = state;

  const require = createRequire(filename);
  const moduleName = path.node.source.value;

  let modulePaths: string[] = [];
  let returnComponentsAsBlocked = false;

  let isNodeModule: boolean;
  let isBlocked: boolean | undefined;
  let isAllowed: boolean | undefined;

  try {
    modulePaths = [require.resolve(moduleName)];
    isNodeModule = modulePaths[0].includes("node_modules");
  } catch {
    /**
     * Hello user! If your are reading this then your probably wondering why your exotic import isn't working.
     *
     * You might be using module-alias, ts-config paths, or importing a folder without an index.{js,jsx,ts,tsx} file
     * either way your doing something that doesn't follow the rules of a normal require statement.
     *
     * Because there's many scenarios, we simply try
     *  - Check if its a file
     *  - Check if its a directory that has an platform specific index files
     *  - Scan the allow/block lists and try and work out if this is a module alias
     */
    const guessAtPath = join(dirname(filename), moduleName);
    const existsOnFileSystem = existsSync(guessAtPath);

    if (existsOnFileSystem) {
      if (lstatSync(guessAtPath).isFile()) {
        // Not sure why require would fail if this was a file, but lets account for it anyway
        isNodeModule = false;
        modulePaths.push(guessAtPath);
      } else {
        isNodeModule = false;

        for (const file of readdirSync(guessAtPath)) {
          if (allowedIndexFiles.includes(basename(file))) {
            modulePaths.push(join(guessAtPath, file));
          }
        }
      }
    } else {
      isNodeModule = true;
      isBlocked = micromatch.isMatch(moduleName, blockModuleTransform);
      isAllowed =
        allowModuleTransform === "*"
          ? true
          : micromatch.isMatch(moduleName, allowModuleTransform);
    }
  }

  if (isNodeModule) {
    isBlocked ??= micromatch.isMatch(moduleName, blockModuleTransform);
    isAllowed ??= micromatch.isMatch(moduleName, allowModuleTransform);
    returnComponentsAsBlocked = isBlocked || !isAllowed;
  } else {
    const normalizedAllowRelativeModules = !Array.isArray(allowRelativeModules)
      ? []
      : allowRelativeModules.map((modulePath) =>
          resolve(cwd, modulePath).split(sep).join(posix.sep)
        );

    const isNotAllowedRelative = !modulePaths.some((modulePath) => {
      /**
       * This is my naive way to get path matching working on Windows.
       * Basically I turn it into a posix path which seems to work fine
       *
       * If you are a windows user and understand micromatch, can you please send a PR
       * to do this the proper way
       */
      const posixModulePath = modulePath.split(sep).join(posix.sep);
      return micromatch.isMatch(
        posixModulePath,
        normalizedAllowRelativeModules
      );
    });

    returnComponentsAsBlocked = isNotAllowedRelative;
  }

  if (!returnComponentsAsBlocked) {
    return [];
  }

  // Only return component names
  return path.node.specifiers.flatMap((specifier) => {
    const name = specifier.local.name;

    if (name[0] === name[0].toUpperCase()) {
      return [name];
    }

    return [];
  });
}
