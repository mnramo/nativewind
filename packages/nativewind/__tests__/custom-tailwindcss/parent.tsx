import { Text, View } from "react-native";
import { render } from "@testing-library/react-native";
import { TestProvider } from "../tailwindcss/runner";
import { StyledComponent } from "../../src";
import { tailwindRunner } from "../tailwindcss/runner";
import { PARENT_HOVER } from "../../src/utils/selector";

tailwindRunner("Custom - Parent", [
  [
    "parent:text-white",
    {
      styles: {
        "parent:text-white.children@0": {
          color: "#fff",
        },
      },
      childClasses: {
        "parent:text-white": ["parent:text-white.children"],
      },
      atRules: {
        "parent:text-white.children": [[["selector", "(> *)"]]],
      },
    },
  ],
]);

tailwindRunner("Custom - Parent State Variants", [
  [
    "parent-hover:text-white",
    {
      styles: {
        "parent-hover:text-white": {
          color: "#fff",
        },
      },
      masks: {
        "parent-hover:text-white": PARENT_HOVER,
      },
    },
  ],
]);

describe("Custom - Parent Variant Snapshots", () => {
  test("parent:text-white", () => {
    const tree = render(
      <TestProvider>
        <StyledComponent component={View} className="parent:text-white">
          <Text>A</Text>
          <Text>B</Text>
        </StyledComponent>
      </TestProvider>
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
