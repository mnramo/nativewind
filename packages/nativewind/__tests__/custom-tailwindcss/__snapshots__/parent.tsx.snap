// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Custom - Parent Variant Snapshots parent:text-white 1`] = `
<View>
  <Text
    style={
      [
        {
          "color": "#fff",
        },
      ]
    }
  >
    A
  </Text>
  <Text
    style={
      [
        {
          "color": "#fff",
        },
      ]
    }
  >
    B
  </Text>
</View>
`;
